File: emulator/harness/fuzzware_harness/native/ufuzz_adapter/global_var.cc
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>
#include <cstdint>

// Assuming uc_mem_read_offset_one_byte is a function that reads one byte from a given address
// and returns its value. You would need to define this function based on your specific use case.
// For the sake of this example, I'm defining a placeholder function.
uint8_t uc_mem_read_offset_one_byte(void* uc, uintptr_t address) {
    // Placeholder implementation
    // Replace with actual memory read logic
    return *(reinterpret_cast<uint8_t*>(address));
}

class GlobalVar {
private:
    uintptr_t address;
    uint8_t value;
    uint8_t previous_value;

public:
    GlobalVar(void* uc, const std::string& addressStr) {
        address = std::stoul(addressStr, nullptr, 16);
        value = uc_mem_read_offset_one_byte(uc, address);
        previous_value = value;
    }

    bool value_changed(void* uc) {
        value = uc_mem_read_offset_one_byte(uc, address);
        bool ret_val = (value != previous_value);
        previous_value = value;
        return ret_val;
    }
};

std::vector<int> find_longest_continuous_segment(const std::set<int>& data) {
    std::vector<int> keys(data.begin(), data.end());
    std::vector<std::vector<int>> segments;
    std::vector<int> current_segment;

    if (!keys.empty()) {
        std::sort(keys.begin(), keys.end());
        current_segment.push_back(keys[0]);

        for (size_t i = 1; i < keys.size(); ++i) {
            if (keys[i] == keys[i - 1] + 1) {
                current_segment.push_back(keys[i]);
            } else {
                segments.push_back(current_segment);
                current_segment = {keys[i]};
            }
        }

        segments.push_back(current_segment);
    }

    std::vector<int> longest_segment;
    for (const auto& segment : segments) {
        if (segment.size() > longest_segment.size()) {
            longest_segment = segment;
        }
    }

    return longest_segment;
}

int main() {
    std::set<int> data = {1, 2, 3, 5, 6, 7, 8, 9, 10};
    std::vector<int> longest_segment = find_longest_continuous_segment(data);

    for (int num : longest_segment) {
        std::cout << num << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: emulator/harness/fuzzware_harness/uFuzzAdapterPython/global_var.py
# usage: GlobalVar class is used to track the global variable

from ...utils import (
    uc_mem_read_offset_one_byte
)
class GlobalVar:
    def __init__(self, uc, address):
        self.address = int(address, 16)
        self.value = uc_mem_read_offset_one_byte(uc, self.address)
        self.previous_value = self.value

    def value_changed(self, uc):
        self.value = uc_mem_read_offset_one_byte(uc, self.address)
        ret_val = False
        if self.value == self.previous_value:
            ret_val = False
        else:
            ret_val = True
        self.previous_value = self.value
        return ret_val

def find_longest_continuous_segment(data):
    def find_continuous_segments(keys):
        if not keys:
            return []

        # 将集合的元素按升序排序
        sorted_keys = sorted(keys)

        # 查找连续的地址段
        segments = []
        current_segment = [sorted_keys[0]]

        for i in range(1, len(sorted_keys)):
            if sorted_keys[i] == sorted_keys[i - 1] + 1:
                current_segment.append(sorted_keys[i])
            else:
                segments.append(current_segment)
                current_segment = [sorted_keys[i]]

        segments.append(current_segment)
        return segments

    # 提取集合的元素列表
    keys = list(data)

    # 查找连续的地址段
    continuous_segments = find_continuous_segments(keys)
    print(continuous_segments)
    # 找到最长的连续地址段
    longest_segment = max(continuous_segments, key=len, default=[])

    return longest_segment

if __name__ == '__main__':
    data = {1, 2, 3, 5, 6, 7, 8, 9, 10}
    print(find_longest_continuous_segment(data))